src/ffi/zlib_rs.rs

```diff
@@ -0,0 +1,164 @@
+//! Implementation for `zlib_rs` rust backend.
+
+use std::fmt;
+
+use ::zlib_rs::{DeflateFlush, InflateError, InflateFlush};
+
+pub const MZ_NO_FLUSH: isize = DeflateFlush::NoFlush as isize;
+pub const MZ_PARTIAL_FLUSH: isize = DeflateFlush::PartialFlush as isize;
+pub const MZ_SYNC_FLUSH: isize = DeflateFlush::SyncFlush as isize;
+pub const MZ_FULL_FLUSH: isize = DeflateFlush::FullFlush as isize;
+pub const MZ_FINISH: isize = DeflateFlush::Finish as isize;
+
+pub const MZ_DEFAULT_WINDOW_BITS: core::ffi::c_int = 15;
+
+use super::*;
+
+impl From<::zlib_rs::Status> for crate::mem::Status {
+    fn from(value: ::zlib_rs::Status) -> Self {
+        match value {
+            ::zlib_rs::Status::Ok => crate::mem::Status::Ok,
+            ::zlib_rs::Status::BufError => crate::mem::Status::BufError,
+            ::zlib_rs::Status::StreamEnd => crate::mem::Status::StreamEnd,
+        }
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct ErrorMessage(Option<&'static str>);
+
+impl ErrorMessage {
+    pub fn get(&self) -> Option<&str> {
+        self.0
+    }
+}
+
+pub struct Inflate {
+    pub(crate) inner: ::zlib_rs::Inflate,
+}
+
+impl fmt::Debug for Inflate {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(
+            f,
+            "zlib_rs inflate internal state. total_in: {}, total_out: {}",
+            self.inner.total_in(),
+            self.inner.total_out(),
+        )
+    }
+}
+
+impl From<FlushDecompress> for DeflateFlush {
+    fn from(value: FlushDecompress) -> Self {
+        match value {
+            FlushDecompress::None => Self::NoFlush,
+            FlushDecompress::Sync => Self::SyncFlush,
+            FlushDecompress::Finish => Self::Finish,
+        }
+    }
+}
+
+impl InflateBackend for Inflate {
+    fn make(zlib_header: bool, window_bits: u8) -> Self {
+        Inflate {
+            inner: ::zlib_rs::Inflate::new(zlib_header, window_bits),
```

Comment:
I wonder if there are any benefits to have a pure-Rust implementation of this. Maybe it's less allocations?
In any case, I suppose it will have to allocate internal buffers.
This is just me writing down a minor thought, it's not really relevant.